src/main/modules/s2s/index.ts

```
@@ -4,59 +4,78 @@ import config from 'config';
-                  });
-            })
-            .then(response => response.text())
-            .then((token: string) => {
-              resolve(token);
+      try {
+        // Try to get token from Redis first
+        let serviceToken: string | null = await redisClient?.get(s2sKey);
+
+        if (!serviceToken) {
+          const { otp } = TOTP.generate(s2sSecret);
+
+          const params = {
+            microservice,
+            oneTimePassword: otp,
+          };
+
+          const request = new Promise<string>((resolve, reject) => {
+            fetch(`${s2sUrl}/lease`, {
+              method: 'POST',
+              headers: {
+                'Content-Type': 'application/json',
+              },
+              body: JSON.stringify(params),
             })
-            .catch((err: string) => {
-              this.logger.error(err);
-              reject(err);
-            });
-        });
-
-        try {
-          req.app.locals.serviceToken = (await request) as string;
-        } catch (error) {
-          this.logger.error('S2S ERROR', error.message);
+              .then(response => {
+                return response.status === 200
+                  ? response
+                  : response.text().then(text => {
+                      this.logger.error('FETCH ERR: ', response.status, text);
+                      return Promise.reject(new Error(text));
+                    });
+              })
+              .then(response => response.text())
+              .then((token: string) => {
+                resolve(token);
+              })
+              .catch((err: string) => {
+                this.logger.error(err);
+                reject(err);
+              });
+          });
+
+          try {
+            serviceToken = await request;
+            // Store token in Redis with expiry so it is automatically removed after TTL
+            await redisClient?.set(s2sKey, serviceToken, 'EX', s2sTtl);
```

How about splitting these two statements out into separate try/catch blocks? A failure to store the token in Redis will just result in a slight result in efficiency, but the app will carry on working. However, a failure to get a new s2s token is more significant and in my opinion it would be better to fail fast here with a specific log error message.